File: playback.py
"""
Spotify playback controls — background mode.
"""

import sys

from lib import (
    read_input,
    write_response,
    background,
    error,
    need_setup_response,
    is_authed,
    api,
    strip_keyword,
    fmt_artists,
)


def ensure_device_id(ctx):
    """
    Resolve an active Spotify device id, falling back to the first
    available one if no device is currently active.

    Spotify's /me/player/play endpoint returns 404 NO_ACTIVE_DEVICE
    when nothing is playing — even if the user has Spotify open. This
    helper sidesteps that by picking a device explicitly.
    """
    try:
        cur = api(ctx, "GET", "/me/player")
        if cur and (cur.get("device") or {}).get("id"):
            return cur["device"]["id"]
    except RuntimeError:
        pass
    try:
        data = api(ctx, "GET", "/me/player/devices") or {}
    except RuntimeError:
        return None
    items = data.get("devices") or []
    if not items:
        return None
    for d in items:
        if d.get("is_active"):
            return d.get("id")
    for d in items:
        if not d.get("is_restricted"):
            return d.get("id")
    return items[0].get("id")


def _no_device_error():
    return error(
        "No active Spotify device",
        details="Open Spotify on a phone, computer, or speaker first. `sp d` shows what Yoki sees.",
    )


def _is_no_device_error(e):
    s = str(e).lower()
    return "no active" in s or "no_active_device" in s or "404" in s


def detect_command(input_command, query):
    if input_command in {"play", "pause", "next", "prev", "vol", "shuffle", "like", "unlike"}:
        return input_command
    low = (query or "").strip().lower()
    for kw, name in [
        ("pause", "pause"), ("stop", "pause"),
        ("next", "next"), ("skip", "next"), ("n", "next"),
        ("prev", "prev"), ("previous", "prev"), ("back", "prev"), ("p", "prev"),
        ("vol", "vol"), ("volume", "vol"), ("v", "vol"),
        ("shuffle", "shuffle"), ("sh", "shuffle"),
        ("unlike", "unlike"), ("unfav", "unlike"),
        ("like", "like"), ("love", "like"), ("fav", "like"),
        ("play", "play"),
    ]:
        if low == kw or low.startswith(kw + " "):
            return name
    return "play"


def _is_spotify_uri(s):
    """Check if a string is a Spotify URI (spotify:track:xxx) or URL
    (https://open.spotify.com/...). These should be played directly via
    the context_uris / uris endpoint, not searched."""
    return s.startswith("spotify:") or s.startswith("https://open.spotify.com/")


def cmd_play(ctx, query):
    arg = strip_keyword(query, "play")

    device_id = ensure_device_id(ctx)
    if device_id is None:
        return _no_device_error()

    if not arg:
        try:
            api(ctx, "PUT", "/me/player/play", params={"device_id": device_id})
            return background("Resumed")
        except RuntimeError as e:
            if _is_no_device_error(e):
                return _no_device_error()
            return error("Play failed", details=str(e))

    # Direct Spotify URI/URL — play without searching. Supports all
    # resource types: track, album, playlist, artist, show, episode.
    # Track URIs go into "uris" (single-track queue), everything else
    # goes into "context_uris" (play the whole collection).
    if _is_spotify_uri(arg):
        is_track = ":track:" in arg or "/track/" in arg
        body = {"uris": [arg]} if is_track else {"context_uris": [arg]}
        try:
            api(ctx, "PUT", "/me/player/play", body=body, params={"device_id": device_id})
        except RuntimeError as e:
            if _is_no_device_error(e):
                return _no_device_error()
            return error("Play failed", details=str(e))
        return background(f"Playing {arg.split(':')[-1][:20]}...")

    # Text query — search for a track and play the first result.
    try:
        res = api(ctx, "GET", "/search", params={"q": arg, "type": "track", "limit": 1})
    except RuntimeError as e:
        return error("Search failed", details=str(e))
    items = ((res or {}).get("tracks") or {}).get("items") or []
    if not items:
        return background(f"No tracks found for: {arg}")
    track = items[0]
    uri = track["uri"]
    try:
        api(ctx, "PUT", "/me/player/play", body={"uris": [uri]}, params={"device_id": device_id})
    except RuntimeError as e:
        if _is_no_device_error(e):
            return _no_device_error()
        return error("Play failed", details=str(e))
    return background(
        f"{track['name']} - {fmt_artists(track.get('artists'))}",
        notif={
            "title": "Now playing",
            "body": f"{track['name']} - {fmt_artists(track.get('artists'))}",
        },
    )


def cmd_pause(ctx):
    device_id = ensure_device_id(ctx)
    if device_id is None:
        return _no_device_error()
    try:
        api(ctx, "PUT", "/me/player/pause", params={"device_id": device_id})
    except RuntimeError as e:
        if _is_no_device_error(e):
            return _no_device_error()
        return error("Pause failed", details=str(e))
    return background("Paused")


def cmd_next(ctx):
    device_id = ensure_device_id(ctx)
    if device_id is None:
        return _no_device_error()
    try:
        api(ctx, "POST", "/me/player/next", params={"device_id": device_id})
    except RuntimeError as e:
        if _is_no_device_error(e):
            return _no_device_error()
        return error("Next failed", details=str(e))
    return background("Next track")


def cmd_prev(ctx):
    device_id = ensure_device_id(ctx)
    if device_id is None:
        return _no_device_error()
    try:
        api(ctx, "POST", "/me/player/previous", params={"device_id": device_id})
    except RuntimeError as e:
        if _is_no_device_error(e):
            return _no_device_error()
        return error("Prev failed", details=str(e))
    return background("Previous track")


def cmd_vol(ctx, query):
    arg = strip_keyword(query, "vol", "volume", "v")
    try:
        level = int(arg)
    except (TypeError, ValueError):
        return error("Volume must be 0-100", details=f"Got: {arg!r}")
    level = max(0, min(100, level))
    device_id = ensure_device_id(ctx)
    if device_id is None:
        return _no_device_error()
    try:
        api(ctx, "PUT", "/me/player/volume", params={"volume_percent": level, "device_id": device_id})
    except RuntimeError as e:
        if _is_no_device_error(e):
            return _no_device_error()
        return error("Volume failed", details=str(e))
    bars = "#" * (level // 5) + "-" * (20 - level // 5)
    return background(f"vol {bars} {level}%")


def _current_track_id(ctx):
    """Return the currently-playing track id, or None."""
    cur = api(ctx, "GET", "/me/player/currently-playing")
    if not cur:
        return None
    item = cur.get("item") or {}
    return item.get("id")


def cmd_like(ctx, force_state=None):
    """Toggle like for the current track. force_state=True/False forces add/remove."""
    try:
        track_id = _current_track_id(ctx)
    except RuntimeError as e:
        return error("Like failed", details=str(e))
    if not track_id:
        return background("Nothing playing")

    try:
        contains = api(ctx, "GET", "/me/tracks/contains", params={"ids": track_id})
    except RuntimeError as e:
        # Most likely missing user-library-read scope.
        return error("Like check failed", details=str(e) + " — try `sp setup` to refresh scopes")
    is_liked = bool(contains and contains[0])

    target_liked = (not is_liked) if force_state is None else force_state
    try:
        if target_liked:
            api(ctx, "PUT", "/me/tracks", params={"ids": track_id})
            return background(
                "Added to Liked",
                notif={"title": "Spotify", "body": "Saved to Liked Songs"},
            )
        else:
            api(ctx, "DELETE", "/me/tracks", params={"ids": track_id})
            return background("Removed from Liked")
    except RuntimeError as e:
        msg = str(e)
        if "403" in msg or "scope" in msg.lower() or "insufficient" in msg.lower():
            return error(
                "Missing user-library-modify scope",
                details="Run `sp setup` to re-authenticate with the new permission.",
            )
        return error("Like failed", details=msg)


def cmd_shuffle(ctx):
    try:
        cur = api(ctx, "GET", "/me/player")
    except RuntimeError as e:
        return error("Shuffle failed", details=str(e))
    if not cur:
        return background("Nothing playing")
    new_state = not bool(cur.get("shuffle_state"))
    try:
        api(ctx, "PUT", "/me/player/shuffle", params={"state": "true" if new_state else "false"})
    except RuntimeError as e:
        return error("Shuffle failed", details=str(e))
    return background("Shuffle ON" if new_state else "Shuffle OFF")


def main():
    inp = read_input()
    ctx = inp.get("context", {})
    query = inp.get("query", "") or ""
    command_name = inp.get("command", "")

    if not is_authed(ctx):
        write_response(need_setup_response())
        return

    cmd = detect_command(command_name, query)
    try:
        if cmd == "play":
            resp = cmd_play(ctx, query)
        elif cmd == "pause":
            resp = cmd_pause(ctx)
        elif cmd == "next":
            resp = cmd_next(ctx)
        elif cmd == "prev":
            resp = cmd_prev(ctx)
        elif cmd == "vol":
            resp = cmd_vol(ctx, query)
        elif cmd == "shuffle":
            resp = cmd_shuffle(ctx)
        elif cmd == "like":
            resp = cmd_like(ctx, force_state=None)
        elif cmd == "unlike":
            resp = cmd_like(ctx, force_state=False)
        else:
            resp = error(f"Unknown command: {cmd}")
    except RuntimeError as e:
        if str(e) == "not_authenticated":
            resp = need_setup_response()
        else:
            resp = error("Spotify error", details=str(e))

    write_response(resp)


if __name__ == "__main__":
    try:
        main()
    except Exception as e:
        write_response(error("playback crashed: " + str(e)))
        sys.exit(1)

File: playback.js
#!/usr/bin/env node
"use strict";

/**
 * Spotify playback controls — background mode.
 */

const {
  readInput,
  writeResponse,
  background,
  error,
  needSetupResponse,
  isAuthed,
  api,
  stripKeyword,
  fmtArtists,
} = require("./lib");

async function ensureDeviceId(ctx) {
  try {
    const cur = await api(ctx, "GET", "/me/player");
    if (cur && (cur.device || {}).id) return cur.device.id;
  } catch (_) {}
  let data;
  try {
    data = (await api(ctx, "GET", "/me/player/devices")) || {};
  } catch (_) {
    return null;
  }
  const items = data.devices || [];
  if (!items.length) return null;
  for (const d of items) {
    if (d.is_active) return d.id;
  }
  for (const d of items) {
    if (!d.is_restricted) return d.id;
  }
  return items[0].id;
}

function noDeviceError() {
  return error(
    "No active Spotify device",
    "Open Spotify on a phone, computer, or speaker first. `sp d` shows what Yoki sees."
  );
}

function isNoDeviceError(e) {
  const s = (e.message || "").toLowerCase();
  return s.includes("no active") || s.includes("no_active_device") || s.includes("404");
}

function detectCommand(inputCommand, query) {
  if (["play", "pause", "next", "prev", "vol", "shuffle", "like", "unlike"].indexOf(inputCommand) !== -1) {
    return inputCommand;
  }
  const low = (query || "").trim().toLowerCase();
  const mapping = [
    ["pause", "pause"], ["stop", "pause"],
    ["next", "next"], ["skip", "next"], ["n", "next"],
    ["prev", "prev"], ["previous", "prev"], ["back", "prev"], ["p", "prev"],
    ["vol", "vol"], ["volume", "vol"], ["v", "vol"],
    ["shuffle", "shuffle"], ["sh", "shuffle"],
    ["unlike", "unlike"], ["unfav", "unlike"],
    ["like", "like"], ["love", "like"], ["fav", "like"],
    ["play", "play"],
  ];
  for (const [kw, name] of mapping) {
    if (low === kw || low.startsWith(kw + " ")) return name;
  }
  return "play";
}

function isSpotifyUri(s) {
  return s.startsWith("spotify:") || s.startsWith("https://open.spotify.com/");
}

async function cmdPlay(ctx, query) {
  const arg = stripKeyword(query, "play");
  const deviceId = await ensureDeviceId(ctx);
  if (deviceId == null) return noDeviceError();

  if (!arg) {
    try {
      await api(ctx, "PUT", "/me/player/play", null, { device_id: deviceId });
      return background("Resumed");
    } catch (e) {
      if (isNoDeviceError(e)) return noDeviceError();
      return error("Play failed", e.message);
    }
  }

  if (isSpotifyUri(arg)) {
    const isTrack = arg.includes(":track:") || arg.includes("/track/");
    const body = isTrack ? { uris: [arg] } : { context_uri: arg };
    try {
      await api(ctx, "PUT", "/me/player/play", body, { device_id: deviceId });
    } catch (e) {
      if (isNoDeviceError(e)) return noDeviceError();
      return error("Play failed", e.message);
    }
    return background("Playing " + arg.split(":").pop().slice(0, 20) + "...");
  }

  // Text query — search for a track and play the first result
  let res;
  try {
    res = await api(ctx, "GET", "/search", null, { q: arg, type: "track", limit: 1 });
  } catch (e) {
    return error("Search failed", e.message);
  }
  const items = ((res || {}).tracks || {}).items || [];
  if (!items.length) return background("No tracks found for: " + arg);

  const track = items[0];
  const uri = track.uri;
  try {
    await api(ctx, "PUT", "/me/player/play", { uris: [uri] }, { device_id: deviceId });
  } catch (e) {
    if (isNoDeviceError(e)) return noDeviceError();
    return error("Play failed", e.message);
  }
  const desc = track.name + " - " + fmtArtists(track.artists);
  return background(desc, { title: "Now playing", body: desc });
}

async function cmdPause(ctx) {
  const deviceId = await ensureDeviceId(ctx);
  if (deviceId == null) return noDeviceError();
  try {
    await api(ctx, "PUT", "/me/player/pause", null, { device_id: deviceId });
  } catch (e) {
    if (isNoDeviceError(e)) return noDeviceError();
    return error("Pause failed", e.message);
  }
  return background("Paused");
}

async function cmdNext(ctx) {
  const deviceId = await ensureDeviceId(ctx);
  if (deviceId == null) return noDeviceError();
  try {
    await api(ctx, "POST", "/me/player/next", null, { device_id: deviceId });
  } catch (e) {
    if (isNoDeviceError(e)) return noDeviceError();
    return error("Next failed", e.message);
  }
  return background("Next track");
}

async function cmdPrev(ctx) {
  const deviceId = await ensureDeviceId(ctx);
  if (deviceId == null) return noDeviceError();
  try {
    await api(ctx, "POST", "/me/player/previous", null, { device_id: deviceId });
  } catch (e) {
    if (isNoDeviceError(e)) return noDeviceError();
    return error("Prev failed", e.message);
  }
  return background("Previous track");
}

async function cmdVol(ctx, query) {
  const arg = stripKeyword(query, "vol", "volume", "v");
  const level = parseInt(arg, 10);
  if (isNaN(level)) {
    return error("Volume must be 0-100", "Got: " + JSON.stringify(arg));
  }
  const clamped = Math.max(0, Math.min(100, level));
  const deviceId = await ensureDeviceId(ctx);
  if (deviceId == null) return noDeviceError();
  try {
    await api(ctx, "PUT", "/me/player/volume", null, { volume_percent: clamped, device_id: deviceId });
  } catch (e) {
    if (isNoDeviceError(e)) return noDeviceError();
    return error("Volume failed", e.message);
  }
  const bars = "#".repeat(Math.floor(clamped / 5)) + "-".repeat(20 - Math.floor(clamped / 5));
  return background("vol " + bars + " " + clamped + "%");
}

async function currentTrackId(ctx) {
  const cur = await api(ctx, "GET", "/me/player/currently-playing");
  if (!cur) return null;
  return (cur.item || {}).id || null;
}

async function cmdLike(ctx, forceState) {
  let trackId;
  try {
    trackId = await currentTrackId(ctx);
  } catch (e) {
    return error("Like failed", e.message);
  }
  if (!trackId) return background("Nothing playing");

  let contains;
  try {
    contains = await api(ctx, "GET", "/me/tracks/contains", null, { ids: trackId });
  } catch (e) {
    return error("Like check failed", e.message + " \u2014 try `sp setup` to refresh scopes");
  }
  const isLiked = !!(contains && contains[0]);
  const targetLiked = forceState === undefined ? !isLiked : forceState;

  try {
    if (targetLiked) {
      await api(ctx, "PUT", "/me/tracks", null, { ids: trackId });
      return background("Added to Liked", { title: "Spotify", body: "Saved to Liked Songs" });
    } else {
      await api(ctx, "DELETE", "/me/tracks", null, { ids: trackId });
      return background("Removed from Liked");
    }
  } catch (e) {
    const msg = e.message || "";
    if (msg.includes("403") || msg.toLowerCase().includes("scope") || msg.toLowerCase().includes("insufficient")) {
      return error(
        "Missing user-library-modify scope",
        "Run `sp setup` to re-authenticate with the new permission."
      );
    }
    return error("Like failed", msg);
  }
}

async function cmdShuffle(ctx) {
  let cur;
  try {
    cur = await api(ctx, "GET", "/me/player");
  } catch (e) {
    return error("Shuffle failed", e.message);
  }
  if (!cur) return background("Nothing playing");
  const newState = !cur.shuffle_state;
  try {
    await api(ctx, "PUT", "/me/player/shuffle", null, { state: newState ? "true" : "false" });
  } catch (e) {
    return error("Shuffle failed", e.message);
  }
  return background(newState ? "Shuffle ON" : "Shuffle OFF");
}

async function main() {
  const inp = await readInput();
  const ctx = inp.context || {};
  const query = inp.query || "";
  const commandName = inp.command || "";

  if (!isAuthed(ctx)) {
    writeResponse(needSetupResponse());
    return;
  }

  const cmd = detectCommand(commandName, query);
  let resp;
  try {
    if (cmd === "play") resp = await cmdPlay(ctx, query);
    else if (cmd === "pause") resp = await cmdPause(ctx);
    else if (cmd === "next") resp = await cmdNext(ctx);
    else if (cmd === "prev") resp = await cmdPrev(ctx);
    else if (cmd === "vol") resp = await cmdVol(ctx, query);
    else if (cmd === "shuffle") resp = await cmdShuffle(ctx);
    else if (cmd === "like") resp = await cmdLike(ctx);
    else if (cmd === "unlike") resp = await cmdLike(ctx, false);
    else resp = error("Unknown command: " + cmd);
  } catch (e) {
    if (e.message === "not_authenticated") {
      resp = needSetupResponse();
    } else {
      resp = error("Spotify error", e.message);
    }
  }

  writeResponse(resp);
}

main().catch(function (e) {
  writeResponse(error("playback crashed: " + e.message));
  process.exit(1);
});
